developer/frameworks/articles/dependency-injection/02-cdi-dependency-injection.markdown

```diff
@@ -0,0 +1,142 @@
+    | use CDI via
+    | [OSGi CDI Integration](/docs/7-2/frameworks/-/knowledge_base/f/osgi-cdi-integration). 
+
+    | **Note:**
+    | [@product@ exports](/docs/7-2/reference/-/knowledge_base/r/third-party-packages-product-exports)
+    | the `portlet-api` and `cdi-api` packages. Projects generated using 
+    | Liferay project templates typically include them as transitive 
+    | dependencies. If you must explicitly depend on the portlet API and CDI
+    | artifacts, add them as `compileOnly` (Gradle) or `provided` (Maven)
+    | dependencies. 
+
+2.  If your portlet WAR project isn't a Bean Portlet, add this
+    `src/main/webapp/WEB-INF/beans.xml` file to it. This file informs CDI to
+    scan the project for CDI annotations. 
+
+```xml 
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://xmlns.jcp.org/xml/ns/javaee" bean-discovery-mode="all" version="1.2"
+	   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	   xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/beans_1_1.xsd">
+	<!-- This file is necessary in order to inform CDI that scanning should occur for CDI annotations. -->
+</beans>
+```
+
+| **Note:** In a PortletMVC4Spring Portlet, this `bean.xml` replaces the
+| application context file (e.g.,
+| `src/main/webapp/WEB-INF/applicationContext.xml`). 
+
+3.  Add the
+    [`@ApplicationScoped`](https://docs.oracle.com/javaee/7/api/javax/enterprise/context/ApplicationScoped.html)
+    annotation to your portlet class. 
+
+```java 
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class MyPortlet ... {
+    ...
+}
+```
+
+4.  Make sure all concrete classes you want to make injectable have the default 
+    constructor. These classes are now CDI beans. 
+
+5.  Add a scope to each CDI bean. 
+
+| Bean Scope              | Description      |
+| ----------------------- | ---------------- |
+| [`@ApplicationScoped`](https://docs.oracle.com/javaee/7/api/javax/enterprise/context/ApplicationScoped.html) | Shares the bean's state across all users' interactions with the portlet. |
+| [`@Dependent`](https://docs.oracle.com/javaee/7/api/javax/enterprise/context/Dependent.html) | (default scope) Designates the bean to be for the client bean and share the client bean's lifecycle. |
+| [`@PortletRequestScoped`](https://docs.liferay.com/portlet-api/3.0/javadocs/javax/portlet/annotations/PortletRequestScoped.html) | Associates the bean with the portlet request. |
+| [`@PortletSessionScoped`](https://docs.liferay.com/portlet-api/3.0/javadocs/javax/portlet/annotations/PortletSessionScoped.html) | Places the bean in the portlet session. |
+| [`@RenderStateScoped`](https://docs.liferay.com/portlet-api/3.0/javadocs/javax/portlet/annotations/RenderStateScoped.html) | Stores the bean as part of the portlet's render state. **Important:** The bean must implement the `PortletSerializable` interface. |
```

Comment:
@jhinkey @sez11a Liferay 7.2 GA2 will feature [RedirectScoped](https://static.javadoc.io/javax.mvc/javax.mvc-api/1.0-pfd/javax/mvc/RedirectScoped.html) as well. It will be essential for MVCBean portlets, which of course are  bean portlets that use the MVC 1.0 API.